File: src/main/java/ludwig/interpreter/CallableRef.java
package ludwig.interpreter;

import ludwig.model.FunctionNode;
import ludwig.model.Node;
import ludwig.utils.NodeUtils;
import org.pcollections.HashTreePMap;


public class CallableRef implements Callable {
    private final Node<?> function;
    private int argCount;

    public CallableRef(Node<?> function) {
        this.function = function;
        argCount = NodeUtils.INSTANCE.arguments(function).size(); // TODO: Optimize
    }

    @Override
    public Object tail(Object... args) {
        return new Evaluator(HashTreePMap.empty()).tail(function, args);
    }

    @Override
    public boolean isLazy() {
        return (function instanceof FunctionNode) && ((FunctionNode)function).lazy();
    }

    @Override
    public int argCount() {
        return argCount;
    }

    @Override
    public String toString() {
        return "ref " + function;
    }
}

File: src/main/java/ludwig/interpreter/CallableRef.kt
package ludwig.interpreter

import ludwig.model.FunctionNode
import ludwig.model.Node
import ludwig.utils.NodeUtils
import org.pcollections.HashTreePMap


class CallableRef(private val function: Node<*>) : Callable {
    private val argCount: Int

    init {
        argCount = NodeUtils.arguments(function).size // TODO: Optimize
    }

    override fun tail(args: Array<Any?>): Any? {
        return Evaluator(HashTreePMap.empty()).tail(function, args)
    }

    override val isLazy: Boolean
        get() = function is FunctionNode && function.lazy

    override fun argCount(): Int {
        return argCount
    }

    override fun toString(): String {
        return "ref " + function
    }
}
